cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetInstanceService.java

```diff
@@ -107,10 +128,29 @@ public void setAuditPublisher(AuditPublisher auditPublisher) {
    * @throws NotFoundException if the namespace was not found
    * @throws IOException if there is a problem in making an HTTP request to check if the namespace exists.
    */
-  public Collection<DatasetSpecification> list(Id.Namespace namespace) throws Exception {
+  Collection<DatasetSpecification> list(final Id.Namespace namespace) throws Exception {
+    Principal principal = SecurityRequestContext.toPrincipal();
     // Throws NamespaceNotFoundException if the namespace does not exist
     ensureNamespaceExists(namespace);
-    return instanceManager.getAll(namespace);
+    Collection<DatasetSpecification> datasets = instanceManager.getAll(namespace);
+    Collection<DatasetId> datasetIds = Collections2.transform(
+      datasets, new Function<DatasetSpecification, DatasetId>() {
+        @Override
+        public DatasetId apply(DatasetSpecification spec) {
+          return namespace.toEntityId().dataset(spec.getName());
+        }
+      });
+    final Set<DatasetId> authorized =
+      authorizationEnforcementService.filter(Sets.newHashSet(datasetIds), principal);
```

Comment:
Now when I looking at the usage, it's probably better that the `AuthorizationEnforcer` expose the `Predicate` for filter (we have `Predicate` in cdap-api that you can use). With that, you don't have to copy and create new set. With that, this method can simply be: 

```java
final NamespaceId ns = namespace.toEntityId();
final Predicate<EntityId> filter = authEnforcer.createFilter(principal);
Collection<DatasetSpecification> datasets = ....
return new Lists.newArrayList(Iterables.filter(datasets, new Predicate<DatasetSpecification>() {
  @Override
  public boolean apply(DatasetSpecification spec) {
    return filter.apply(ns.dataset(spec.getName()));
  }
}));
```